File: web/src/shared/section-catalog.ts
import type { DocumentType } from "./models";

export const SectionCatalog = {
  DATOS_PACIENTE: "Datos del paciente",
  MOTIVO_CONSULTA: "Motivo de consulta",
  ENFERMEDAD_ACTUAL: "Enfermedad actual",
  ANTECEDENTES_PERSONALES: "Antecedentes personales",
  ANTECEDENTES_FAMILIARES: "Antecedentes familiares",
  HABITOS_PSICOBIOLOGICOS: "Hábitos psicobiológicos",
  EXAMEN_FUNCIONAL: "Examen funcional",
  EXAMEN_FISICO: "Examen físico",
  DIAGNOSTICO: "Diagnóstico",
  IMPRESION_DIAGNOSTICA: "Impresión diagnóstica",
  PLAN: "Plan",
  OBSERVACIONES: "Observaciones",
  CONCLUSIONES: "Conclusiones",
  RECOMENDACIONES: "Recomendaciones",
  DIAS_REPOSO: "Días de reposo indicados",
  INDICACIONES: "Indicaciones",
  ORDENES: "Órdenes",
  RECIPE: "Recipe",
} as const;

export function catalogFor(type: DocumentType): string[] {
  switch (type) {
    case "historiaClinica":
      return [
        SectionCatalog.DATOS_PACIENTE,
        SectionCatalog.MOTIVO_CONSULTA,
        SectionCatalog.ENFERMEDAD_ACTUAL,
        SectionCatalog.ANTECEDENTES_PERSONALES,
        SectionCatalog.ANTECEDENTES_FAMILIARES,
        SectionCatalog.HABITOS_PSICOBIOLOGICOS,
        SectionCatalog.EXAMEN_FUNCIONAL,
        SectionCatalog.EXAMEN_FISICO,
        SectionCatalog.DIAGNOSTICO,
        SectionCatalog.IMPRESION_DIAGNOSTICA,
        SectionCatalog.PLAN,
        SectionCatalog.OBSERVACIONES,
      ];
    case "informe":
      return [
        SectionCatalog.DATOS_PACIENTE,
        SectionCatalog.MOTIVO_CONSULTA,
        SectionCatalog.ENFERMEDAD_ACTUAL,
        SectionCatalog.EXAMEN_FISICO,
        SectionCatalog.DIAGNOSTICO,
        SectionCatalog.CONCLUSIONES,
        SectionCatalog.PLAN,
      ];
    case "reposo":
      return [
        SectionCatalog.DATOS_PACIENTE,
        SectionCatalog.DIAGNOSTICO,
        SectionCatalog.DIAS_REPOSO,
        SectionCatalog.INDICACIONES,
        SectionCatalog.OBSERVACIONES,
      ];
    case "ordenesMedicas":
      return [SectionCatalog.DATOS_PACIENTE, SectionCatalog.ORDENES];
    case "receta":
      return [SectionCatalog.DATOS_PACIENTE, SectionCatalog.RECIPE, SectionCatalog.INDICACIONES];
  }
}

export function defaultSectionsFor(type: DocumentType): string[] {
  switch (type) {
    case "informe":
      // Primera vez: solo núcleo clínico (Conclusiones/Plan desmarcados)
      return [
        SectionCatalog.DATOS_PACIENTE,
        SectionCatalog.MOTIVO_CONSULTA,
        SectionCatalog.ENFERMEDAD_ACTUAL,
        SectionCatalog.EXAMEN_FISICO,
        SectionCatalog.DIAGNOSTICO,
      ];
    case "historiaClinica":
    case "reposo":
    case "ordenesMedicas":
    case "receta":
      return catalogFor(type);
  }
}

function sameSections(a: string[], b: string[]): boolean {
  return a.length === b.length && a.every((s, i) => s === b[i]);
}

/** Catálogo completo del informe (legado: solían venir todas checadas). */
export const INFORME_FULL_CATALOG = catalogFor("informe");

/** Antiguo catálogo completo con "Recomendaciones" (antes del rename a Plan). */
const INFORME_LEGACY_FULL = [
  SectionCatalog.DATOS_PACIENTE,
  SectionCatalog.MOTIVO_CONSULTA,
  SectionCatalog.ENFERMEDAD_ACTUAL,
  SectionCatalog.EXAMEN_FISICO,
  SectionCatalog.DIAGNOSTICO,
  SectionCatalog.CONCLUSIONES,
  "Recomendaciones",
];

export function isLegacyInformeAllChecked(sections: string[]): boolean {
  return sameSections(sections, INFORME_FULL_CATALOG) || sameSections(sections, INFORME_LEGACY_FULL);
}

export function normalizeTemplateSections(type: DocumentType, sections: string[]): string[] {
  const catalog = catalogFor(type);
  const mapped = sections
    .map((s) => (s.trim().toLowerCase() === "recomendaciones" ? SectionCatalog.PLAN : s.trim()))
    .filter((s) => s.length > 0);
  // Conserva secciones personalizadas fuera del catálogo
  const unique = [...new Set(mapped)];
  if (
    !unique.includes(SectionCatalog.DATOS_PACIENTE) &&
    catalog.includes(SectionCatalog.DATOS_PACIENTE)
  ) {
    return [SectionCatalog.DATOS_PACIENTE, ...unique.filter((s) => s !== SectionCatalog.DATOS_PACIENTE)];
  }
  return unique.length ? unique : defaultSectionsFor(type);
}

File: app/src/main/java/com/ceoclinicos/clinicosdoc/model/SectionCatalog.kt
package com.ceoclinicos.clinicosdoc.model

object SectionCatalog {
    const val MOTIVO_CONSULTA = "Motivo de consulta"
    const val ENFERMEDAD_ACTUAL = "Enfermedad actual"
    const val ANTECEDENTES_PERSONALES = "Antecedentes personales"
    const val ANTECEDENTES_FAMILIARES = "Antecedentes familiares"
    const val EXAMEN_FISICO = "Examen físico"
    const val EXAMEN_FUNCIONAL = "Examen funcional"
    const val IMPRESION_DIAGNOSTICA = "Impresión diagnóstica"
    const val PLAN = "Plan"
    const val OBSERVACIONES = "Observaciones"
    const val DATOS_PACIENTE = "Datos del paciente"
    const val MOTIVO_INFORME = "Motivo del informe"
    const val HALLAZGOS_CLINICOS = "Hallazgos clínicos"
    const val CONCLUSIONES = "Conclusiones"
    const val RECOMENDACIONES = "Recomendaciones"
    const val DIAGNOSTICO = "Diagnóstico"
    const val HABITOS_PSICOBIOLOGICOS = "Hábitos psicobiológicos"
    const val DIAS_REPOSO = "Días de reposo indicados"
    const val INDICACIONES = "Indicaciones"
    const val ORDENES = OrdenesMedicasDefaults.SECTION_ORDENES

    val all = listOf(
        MOTIVO_CONSULTA,
        ENFERMEDAD_ACTUAL,
        ANTECEDENTES_PERSONALES,
        ANTECEDENTES_FAMILIARES,
        EXAMEN_FISICO,
        EXAMEN_FUNCIONAL,
        IMPRESION_DIAGNOSTICA,
        PLAN,
        OBSERVACIONES,
        DATOS_PACIENTE,
        MOTIVO_INFORME,
        HALLAZGOS_CLINICOS,
        CONCLUSIONES,
        RECOMENDACIONES,
        DIAGNOSTICO,
        HABITOS_PSICOBIOLOGICOS,
        DIAS_REPOSO,
        INDICACIONES,
        ORDENES,
    )

    /** Secciones disponibles para elegir/ordenar por tipo de documento. */
    fun catalogFor(documentType: DocumentType): List<String> = when (documentType) {
        DocumentType.HISTORIA_CLINICA -> listOf(
            DATOS_PACIENTE,
            MOTIVO_CONSULTA,
            ENFERMEDAD_ACTUAL,
            ANTECEDENTES_PERSONALES,
            ANTECEDENTES_FAMILIARES,
            HABITOS_PSICOBIOLOGICOS,
            EXAMEN_FUNCIONAL,
            EXAMEN_FISICO,
            DIAGNOSTICO,
            IMPRESION_DIAGNOSTICA,
            PLAN,
            OBSERVACIONES,
        )
        DocumentType.INFORME -> listOf(
            DATOS_PACIENTE,
            MOTIVO_CONSULTA,
            ENFERMEDAD_ACTUAL,
            EXAMEN_FISICO,
            DIAGNOSTICO,
            CONCLUSIONES,
            PLAN,
        )
        DocumentType.REPOSO -> listOf(
            DATOS_PACIENTE,
            DIAGNOSTICO,
            DIAS_REPOSO,
            INDICACIONES,
            OBSERVACIONES,
        )
        DocumentType.ORDENES_MEDICAS -> listOf(
            DATOS_PACIENTE,
            ORDENES,
        )
    }

    fun requiresLockedPatientSection(documentType: DocumentType): Boolean =
        catalogFor(documentType).firstOrNull() == DATOS_PACIENTE

    /** Asegura «Datos del paciente» primero; permite secciones personalizadas fuera del catálogo. */
    fun normalizeActive(documentType: DocumentType, sections: List<String>): List<String> {
        val catalog = catalogFor(documentType).toSet()
        val mapped = sections.map { s ->
            if (s.equals(RECOMENDACIONES, ignoreCase = true)) PLAN else s.trim()
        }.filter { it.isNotBlank() }
        val unique = mapped.distinct()
        // Catálogo + personalizadas (no descartar custom)
        val kept = unique.filter { it in catalog || it !in catalog }
        if (!requiresLockedPatientSection(documentType)) return kept
        val withoutLocked = kept.filterNot { it == DATOS_PACIENTE }
        return listOf(DATOS_PACIENTE) + withoutLocked
    }

    /** Orden visual: activas en su orden, luego inactivas según catálogo. */
    fun initialLayoutOrder(documentType: DocumentType, activeSections: List<String>): List<String> {
        val catalog = catalogFor(documentType)
        val active = normalizeActive(documentType, activeSections)
        val inactive = catalog.filter { it !in active.toSet() }
        return active + inactive
    }

    fun activeFromLayout(layoutOrder: List<String>, activeSections: List<String>): List<String> {
        val activeSet = activeSections.toSet()
        return layoutOrder.filter { it in activeSet }
    }

    fun defaultsFor(documentType: DocumentType): List<String> = when (documentType) {
        DocumentType.HISTORIA_CLINICA -> listOf(
            DATOS_PACIENTE,
            MOTIVO_CONSULTA,
            ENFERMEDAD_ACTUAL,
            ANTECEDENTES_PERSONALES,
            ANTECEDENTES_FAMILIARES,
            HABITOS_PSICOBIOLOGICOS,
            EXAMEN_FUNCIONAL,
            EXAMEN_FISICO,
            DIAGNOSTICO,
        )
        // Informe: núcleo clínico por defecto (Conclusiones/Plan opcionales)
        DocumentType.INFORME -> listOf(
            DATOS_PACIENTE,
            MOTIVO_CONSULTA,
            ENFERMEDAD_ACTUAL,
            EXAMEN_FISICO,
            DIAGNOSTICO,
        )
        DocumentType.REPOSO -> listOf(
            DATOS_PACIENTE,
            DIAGNOSTICO,
            DIAS_REPOSO,
            INDICACIONES,
            OBSERVACIONES,
        )
        DocumentType.ORDENES_MEDICAS -> listOf(
            DATOS_PACIENTE,
            ORDENES,
        )
    }
}
